lib/pbench/server/database/models/datasets.py

```diff
@@ -407,14 +406,24 @@ class Dataset(Database.Base):
 
     __tablename__ = "datasets"
 
+    # This dict defines the allowed dataset state transitions through its
+    # lifecycle. We have a set of "-ING" action states while the dataset is
+    # being mutated, and a set of "-ED" states designating the last successful
+    # mutation, from UPLOAD through INDEX and eventually DELETE.
+    #
+    # All "-ING" action states allow self-transitions, which provides for a
+    # failed action to be retried later.
+    #
+    # DELETED does not appear as a key here: we define it for completion, but
+    # (A) if a dataset entered DELETED state it could never leave and (B) on
+    # completion of the dataset delete action the dataset's SQL row has been
+    # deleted so we never actually set it to "DELETED" state.
     transitions = {
-        States.UPLOADING: [States.UPLOADED, States.QUARANTINED],
-        States.UPLOADED: [States.INDEXING, States.QUARANTINED],
-        States.INDEXING: [States.INDEXED, States.QUARANTINED],
-        States.INDEXED: [States.INDEXING, States.DELETING, States.QUARANTINED],
-        States.DELETING: [States.DELETED, States.INDEXED, States.QUARANTINED],
-        # NOTE: DELETED and QUARANTINED do not appear as keys in the table
-        # because they're terminal states that cannot be exited.
+        States.UPLOADING: [States.UPLOADED, States.UPLOADING, States.DELETING],
+        States.UPLOADED: [States.INDEXING, States.DELETING],
+        States.INDEXING: [States.INDEXED, States.INDEXING, States.DELETING],
+        States.INDEXED: [States.INDEXING, States.DELETING],
+        States.DELETING: [States.DELETED, States.INDEXED, States.DELETING],
```

Comment:
The transition from `DELETING` to `INDEXED` is curious.  It seems like, if the deletion failed, then the dataset should stay in the `DELETING` state (i.e., that's what happens for `UPLOADING` and `INDEXING`, right?).  What would cause a dataset to transition from `DELETING` "back" to `INDEXED` (especially if it had never been in `INDEXED` before, e.g., because it transitioned directly from `UPLOADED` to `DELETING`)?

Reply:
It is. "Curious," that is, if not downright odd. I don't remember what "logic" led me to define that transition in the first place, and I didn't try to re-evaluate it in this PR. It may have been a thought that a failed delete puts it back in `INDEXED` ... in which case that thought was terminally flawed because, as you point out, the index may well be gone. (And maybe this is partly because I did the state transition table long before implementing the delete operation.)